src/core/ContactUser.cpp

```diff
@@ -307,4 +381,132 @@ void ContactUser::incomingProtocolSocket(QTcpSocket *socket)
         m_outgoingSocket = 0;
     }
 }
+#else
+void ContactUser::assignConnection(Protocol::Connection *connection)
+{
+    if (connection == m_connection) {
+        BUG() << "Connection is already assigned to this ContactUser";
+        return;
+    }
+
+    if (qobject_cast<ContactUser*>(connection->parent()) && connection->parent() != this) {
+        BUG() << "Connection is already owned by another ContactUser";
+        connection->close();
+        return;
+    }
+
+    connection->setParent(this);
+    bool isOutbound = connection->direction() == Protocol::Connection::ClientSide;
+
+    if (!connection->isConnected()) {
+        BUG() << "Connection assigned to contact but isn't connected; discarding";
+        connection->close();
+        connection->deleteLater();
+        return;
+    }
+
+    if (!connection->hasAuthenticatedAs(Protocol::Connection::HiddenServiceAuth, hostname())) {
+        BUG() << "Connection assigned to contact without matching authentication";
+        connection->close();
+        connection->deleteLater();
+        return;
+    }
+
+    if (m_connection && !m_connection->isConnected()) {
+        qDebug() << "Replacing dead connection with new connection";
+        clearConnection();
+    }
+
+    /* To resolve a race if two contacts try to connect at the same time:
+     *
+     * If the existing connection is in the same direction as the new one,
+     * always use the new one.
+     */
+    if (m_connection && connection->direction() == m_connection->direction()) {
+        qDebug() << "Replacing existing connection with contact because the new one goes the same direction";
+        clearConnection();
+    }
+
+    /* If the existing connection is more than 30 seconds old, measured from
+     * when it was successfully established, it's replaced with the new one.
+     */
+    if (m_connection && m_connection->age() > 30) {
```

Comment:
1. We have an active outbound connection
2. That connection has failed, but we haven't noticed the failure on this end
3. The peer has noticed failure, and makes an inbound connection to repair it
4. The peer loses the `preferOutbound` comparison below

In this case, we would keep the old, actually-dead outbound connection instead of using the new inbound one, until we noticed the failure. This does happen in practice. Always preferring a new connection over a >30 second old connection solves it.

The assumptions behind this are:

1. 30 seconds is long enough to avoid the "race for connection" behavior that `preferOutbound` handles. This is reasonable, because it's twice as long as the authentication timeout.
2. If the peer is making a new connection instead of using the established one, they probably know that it has failed and we don't.

There's not a security risk to letting the peer control this decision (and they already can).